### ISSUE TITLE
HTTP Keep-Alive Connections not handled correctly

### ISSUE THREAD

**Issue by hexaclock** [cherrypy/cheroot#91]
* **I'm submitting a ...**
- [x] bug report
- [ ] feature request
- [ ] question about the decisions made in the repository

* **Do you want to request a *feature* or report a *bug*?**
Bug

* **What is the current behavior?**
When the number of open Keep-Alive sessions exceeds CherryPy's thread count specified by `server.thread_pool`, CherryPy queues all new Keep-Alive connections up to the timeout specified by `server.socket_timeout`.

* **If the current behavior is a bug, please provide the steps to reproduce and if possible a screenshots and logs of the problem. If you can, show us your code.**
```python
# mini_cherrypy_server.py
import cherrypy


class HelloWorld(object):
    @cherrypy.expose
    def index(self):
        print(cherrypy.request.headers)
        return "Hello world!"


if __name__ == '__main__':
    cherrypy.config.update({'server.thread_pool': 5})
    cherrypy.quickstart(HelloWorld())
```

```python
# mini_client.py
import requests
from multiprocessing.dummy import Pool as ThreadPool

urls = ['http://localhost:8080/index'] * 10

def do_request(url):
    sess = requests.Session() # forces a keep-alive session
    resp = sess.get(url)
    print(resp.text)

def main():
    thread_pool = ThreadPool()
    results = thread_pool.map(do_request, urls)
    thread_pool.close()
    thread_pool.join()
    
if __name__ == '__main__':
    main()
```

* **What is the expected behavior?**
Allow threads in the thread pool to answer and process new connections by rotating "sleeping" sessions back into the queue.

* **What is the motivation / use case for changing the behavior?**
Correctly support the HTTP/1.1 spec.

* **Please tell us about your environment:**

- Cheroot version: 6.0.0
- CherryPy version: 13.1.0 (if applicable)
- Python version: 3.5.2 32-bit
- OS: Windows 10
- Browser: [all]

**Comment by webknjaz** [comment]
Thanks, I'll take a look at this when I'll have some time

**Comment by webknjaz** [comment]
@hexaclock I haven't had a chance to look into it, sorry. But likely you're right.
We'll likely need to need smth in https://github.com/cherrypy/cheroot/blob/master/cheroot/workers/threadpool.py, but I'm not sure.
If I understand correctly what you described, TCP connections get lost in worker pool, which thinks that it's one HTTP request per connection allowed. On the other hand, I certainly saw some code implementing HTTP/1.1 pipelining somewhere around HTTP parser code.

From your description it looks like we have to maintain HTTP requests separately from TCP connections (which needs checking).

If you have any other findings, feel free to post (or PR) them here :)

cc: @jaraco

**Comment by hexaclock** [comment]
@webknjaz I put something together. https://github.com/hexaclock/cheroot/pull/1

**Comment by webknjaz** [comment]
@hexaclock hey can you create a PR against this upstream repo instead of your own?

**Comment by jaraco** [comment]
Fixed in #199 and releasing as v8.1.0.
